pam/pam_test/module-transaction-dummy_test.go

```diff
@@ -0,0 +1,512 @@
+
+			require.False(t, tc.skipPut && tc.expectedValue == nil)
+
+			tx := NewModuleTransactionDummy(nil)
+			envList, err := tx.GetEnvList()
+			require.NoError(t, err)
+			require.Len(t, envList, 0)
+
+			if tc.presetValues != nil && !tc.skipPut {
+				for env, value := range tc.presetValues {
+					err := tx.PutEnv(env + "=" + value)
+					require.NoError(t, err)
+				}
+				envList, err = tx.GetEnvList()
+				require.NoError(t, err)
+				require.Equal(t, tc.presetValues, envList)
+			}
+
+			if !tc.skipPut {
+				var env string
+				if tc.value != nil {
+					env = tc.env + "=" + *tc.value
+				} else {
+					env = tc.env
+				}
+				err := tx.PutEnv(env)
+				require.ErrorIs(t, err, tc.expectedPutError)
+
+				if tc.expectedPutError == nil {
+					environment, err := tx.GetEnvList()
+					require.NoError(t, err)
+
+					if tc.value != nil && tc.expectedValue != nil {
+						require.Equal(t, map[string]string{
+							tc.env: *tc.expectedValue}, environment)
+					} else if tc.value != nil {
+						require.Equal(t, map[string]string{tc.env: *tc.value},
+							environment)
+					} else {
+						require.Equal(t, map[string]string{}, environment)
+					}
+				} else {
+					environment, err := tx.GetEnvList()
+					require.NoError(t, err)
+					require.Equal(t, map[string]string{}, environment)
+				}
+			}
+
+			if tc.expectedValue != nil {
+				value := tx.GetEnv(tc.env)
+				require.Equal(t, value, *tc.expectedValue)
+			}
+		})
+	}
+}
+
+func TestSetGetData(t *testing.T) {
+	t.Parallel()
+	t.Cleanup(MaybeDoLeakCheck)
+
+	testCases := map[string]struct {
+		key              string
+		data             any
+		presetData       map[string]any
+		expectedSetError error
+		expectedGetError error
+		expectedData     any
+		skipSet          bool
+		skipGet          bool
+	}{
+		"get missing": {
+			skipSet:          true,
+			key:              "not set",
+			expectedGetError: pam.NoModuleData,
+		},
+
+		"set-get data": {
+			presetData:   map[string]any{"some-data": []any{"hey! That's", true}},
+			key:          "data",
+			data:         []any{"hey! That's", true},
+			expectedData: []any{"hey! That's", true},
+		},
+
+		"set replaces data": {
+			presetData: map[string]any{"some-data": []any{"hey! That's", true}},
+			key:        "some-data",
+			data: ModuleTransactionDummy{
+				Items: map[pam.Item]string{pam.Tty: "yay"},
+				Env:   map[string]string{"foo": "bar"},
+			},
+			expectedData: ModuleTransactionDummy{
+				Items: map[pam.Item]string{pam.Tty: "yay"},
+				Env:   map[string]string{"foo": "bar"},
+			},
+		},
+
+		"set nil removes data": {
+			presetData:       map[string]any{"some-data": []any{"hey! That's", true}},
+			key:              "some-data",
+			data:             nil,
+			expectedGetError: pam.NoModuleData,
+		},
+	}
+
+	for name, tc := range testCases {
+		tc := tc
+		t.Run(name, func(t *testing.T) {
+			t.Parallel()
+			t.Cleanup(MaybeDoLeakCheck)
+
+			require.False(t, tc.skipGet && tc.skipSet)
```

Comment:
${same comment about asserting on test case values}

(I won't repeat this on the other tests, but it applies as well 😂)